### PR TITLE
fix(packagist): Fix url resolving

### LINKED ISSUE #20649: Packagist lookup error
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.153.1

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

15.8.1

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

Since upgrading to 34.149.0 I've started to get AggregateError's on GitLab composer package repository packages. It was working fine in 34.148.0 

There seems to be a difference in API URLs:
* Before: `https://gitlab.mycompany.com/api/v4/group/2/-/packages/composer/mycompany/product-base-requirements$456a5bb82937e83c047c9f6a3de70ed56f7c21cdc6abfa258bbd14b487774ed6.json`
* After: `https://gitlab.mycompany.com/api/v4/group/2/-/packages/composer/api/v4/group/2/-/packages/composer/mycompany/product-base-requirements$456a5bb82937e83c047c9f6a3de70ed56f7c21cdc6abfa258bbd14b487774ed6.json`

You can see `api/v4/group/2/-/packages/composer` is added twice.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: GET https://gitlab.mycompany.com/api/v4/group/2/-/packages/composer/api/v4/group/2/-/packages/composer/mycompany/product-base-requirements$456a5bb82937e83c047c9f6a3de70ed56f7c21cdc6abfa258bbd14b487774ed6.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=121) (repository=mycompany/product)
DEBUG: GET https://packagist.org/p2/mycompany/product-base-requirements.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=227) (repository=mycompany/product)
DEBUG: GET https://packagist.org/p2/mycompany/product-base-requirements~dev.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=228) (repository=mycompany/product)
DEBUG: Aggregate error is thrown (repository=mycompany/product)
       "err": {
         "name": "AggregateError",
         "message": "\n    HTTPError: Response code 404 (Not Found)\n        at Request.<anonymous> (/builds/mycompany/gitlab/renovate-bot/node_modules/got/dist/source/as-promise/index.js:118:42)\n        at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    HTTPError: Response code 404 (Not Found)\n        at Request.<anonymous> (/builds/mycompany/gitlab/renovate-bot/node_modules/got/dist/source/as-promise/index.js:118:42)\n        at processTicksAndRejections (node:internal/process/task_queues:95:5)",
         "stack": "AggregateError: \n    HTTPError: Response code 404 (Not Found)\n        at Request.<anonymous> (/builds/mycompany/gitlab/renovate-bot/node_modules/got/dist/source/as-promise/index.js:118:42)\n        at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    HTTPError: Response code 404 (Not Found)\n        at Request.<anonymous> (/builds/mycompany/gitlab/renovate-bot/node_modules/got/dist/source/as-promise/index.js:118:42)\n        at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at next (/builds/mycompany/gitlab/renovate-bot/node_modules/p-map/index.js:43:14)\n    at /builds/mycompany/gitlab/renovate-bot/node_modules/p-map/index.js:67:7\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)"
       }
DEBUG: Datasource 404 (repository=mycompany/product)
       "datasource": "packagist",
       "packageName": "mycompany/product-base-requirements",
       "url": "https://packagist.org/p2/mycompany/product-base-requirements.json"
DEBUG: Failed to look up packagist dependency mycompany/product-base-requirements (repository=mycompany/product, packageFile=app-manager/app/Requirements/composer.json, dependency=mycompany/product-base-requirements)
DEBUG: Failed to look up packagist dependency mycompany/product-base-requirements (repository=mycompany/product, packageFile=app/Requirements/composer.json, dependency=mycompany/product-base-requirements)
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- `resolveBaseUrl('https://example.com/foo', '/bar') === 'https://example.com/foo/bar'`
- we need Packagist to join it as `https://example.com/bar`

## Context

- Fix: #20649

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — COMMENTED** [review]
i think we should fix the resolve base url function. or use Url class to correctly resolve relative paths

needs a test

**Comment by zharinov** [comment]
Will be able to write test 1–2 hours later

File: lib/modules/datasource/packagist/index.ts [hunk 1]
```diff
@@ -133,7 +133,10 @@ export class PackagistDatasource extends Datasource {
     registryUrl: string,
     registryMeta: RegistryMeta
   ): string | null {
+    const registryHost = parseUrl(registryUrl)?.origin;
+
```
Comment (secustor) [inline lib/modules/datasource/packagist/index.ts:137]:
```suggestion

    if (is.nullOrUndefined(registryHost)) {
      return null;
    }
```
Is doing the same and is IMO more readable.

Comment (zharinov) [reply lib/modules/datasource/packagist/index.ts:137]: Actually, I don't like the fact `parseUrl()` returns nullable result. There are many places we're sure URLs are valid, but still we have to check its result for null.

**Comment by zharinov** [comment]
> i think we should fix the resolve base url function. or use Url class to correctly resolve relative paths

It was intentional behavior of `resolveBaseUrl()`, it doensn't need fixing

**Comment by zharinov** [comment]
In future, I would rename current `parseUrl()` to `safeParseUrl()` and added non-nullable `parseUrl()` that doesn't catch errors thrown. By the way it's not that easy to make `new URL(...)` throw

**Review by secustor — APPROVED** [review]

File: lib/modules/datasource/packagist/index.ts [hunk 2]
```diff
@@ -142,12 +144,12 @@ export class PackagistDatasource extends Datasource {
       if (hash) {
         url = url.replace('%hash%', hash);
       }
-      return resolveBaseUrl(registryUrl, url);
+      return resolveBaseUrl(registryHost, url);
```
Comment (viceice) [inline lib/modules/datasource/packagist/index.ts:147]: i think we should use the new  url constructor with two args in a later pr

**Review by viceice — APPROVED** [review]

**Pull request merged**